lib/engine/game/g_1846.rb

```diff
@@ -227,17 +227,23 @@ def action_processed(action)
       end
 
       def close_corporation(corporation)
+        @log << "#{corporation.name} closes"
+
         hexes.each do |hex|
           hex.tile.cities.each do |city|
-            next unless city.tokened_by?(corporation)
+            next unless city.tokened_by?(corporation) || city.reserved_by?(corporation)
 
             city.tokens.map! { |token| token&.corporation == corporation ? nil : token }
-            city.reservations.delete(corporation.name)
+            city.reservations.delete(corporation)
           end
         end
 
         corporation.spend(corporation.cash, @bank)
-        @log << "#{corporation.name} closes"
+        corporation.trains.each { |t| t.buyable = false }
```

Comment:
how did these trains even show up to get bought?

Reply:
`Engine::Depot`'s `@trains` still has references to the closed corporation's trains, so they are returned by `Engine::Depot#other_trains`

https://github.com/tobymao/18xx/blob/master/lib/engine/depot.rb#L78-L82